### PR TITLE
CMake: Add optional precompiled header support

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61940

This adds a `USE_PRECOMPILED_HEADERS` option to the CMake build which
precompiles `ATen.h` and also `CUDAContext.h` for the cuda library.
After making a change in `native_functions.yaml`, this speeds up compilation
time by around 15% on my machine.

Differential Revision: [D29988775](https://our.internmc.facebook.com/intern/diff/D29988775)

### CODE REVIEW

**Comment by peterbell10** [comment]
@malfet PTAL.

**Review by malfet — APPROVED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -241,6 +241,11 @@ option(USE_OBSERVERS "Use observers module." OFF)
 option(USE_OPENCL "Use OpenCL" OFF)
 option(USE_OPENCV "Use OpenCV" OFF)
 option(USE_OPENMP "Use OpenMP for parallel code" ON)
+option(USE_PRECOMPILED_HEADERS "Use pre-compiled headers to accelerate build. Requires cmake >= 3.16." OFF)
+if(USE_PRECOMPILED_HEADERS AND (CMAKE_VERSION VERSION_LESS "3.16"))
+  message(FATAL_ERROR "Precompiled headers require cmake >= 3.16")
+endif()
```
Comment (malfet) [inline CMakeLists.txt:247]: Is there a reason not to turn it on by default?
```suggestion
cmake_dependent_option(USE_PRECOMPILED_HEADERS "Use pre-compiled headers to accelerate build. Requires cmake >= 3.16." ON "CMAKE_VERSION VERSION_LESS 3.16" OFF)
```

**Comment by malfet** [comment]
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29988775).

File: CMakeLists.txt (hunk 1 above)
Comment (peterbell10) [reply CMakeLists.txt:247]: I was just being conservative. There is an issue where the pch gets included everywhere, so it's possible for code to compile with PCH enabled but not without, and vice versa. It's also not necessarily a win on all machines, for example on a 32-core machine I see no benefit in overall compile times.

**Pull request closed without merging**